api/clusterlink.net/v1alpha1/operator_types.go

```diff
@@ -0,0 +1,122 @@
+	IngressTypeNodePort IngressType = "NodePort"
+	// IngressTypeLB indicates that the deployment operator created an external ingress of type LoadBalancer.
+	IngressTypeLB IngressType = "LoadBalancer"
+)
+
+// OperatorSpec defines the desired state of a ClusterLink Operator.
+type OperatorSpec struct {
+	DataPlane DataPlaneSpec `json:"dataplane,omitempty"`
+	Ingress   IngressSpec   `json:"ingress,omitempty"`
+
+	// +kubebuilder:validation:Enum=trace;debug;info;warning;error;fatal
+	// +kubebuilder:default=info
+	// LogLevel define the ClusterLink components log level.
+	LogLevel string `json:"logLevel,omitempty"`
+
+	// +kubebuilder:default="ghcr.io/clusterlink-net"
+	// ContainerRegistry is the container registry to pull the ClusterLink project images.
+	ContainerRegistry string `json:"containerRegistry,omitempty"`
+	// +kubebuilder:default="latest"
+	// ImageTag represents the version of the ClusterLink project images.
+	ImageTag string `json:"imageTag,omitempty"`
+}
+
+// OperatorStatus defines the observed state of ClusterlLink Operator.
+type OperatorStatus struct {
+	Controlplane ComponentStatus `json:"controlplane,omitempty"`
+	Dataplane    ComponentStatus `json:"dataplane,omitempty"`
+}
+
+// ComponentStatus defines the status of component in ClusterLink.
+type ComponentStatus struct {
+	// +kubebuilder:validation:Enum=none;ready;error;pending
+	// +kubebuilder:default=none
+	Status  StatusType `json:"status,omitempty"`
+	Message string     `json:"message,omitempty"`
+}
+
+// DataPlaneSpec defines the desired state of the dataplane components in ClusterLink.
+type DataPlaneSpec struct {
+	// +kubebuilder:validation:Enum=envoy;go
+	// +kubebuilder:default=envoy
+	// Type represents the dataplane type. Supports values "go" and "envoy".
+	Type string `json:"type,omitempty"`
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=10
+	// +kubebuilder:default=1
+	// Replicas represents the number of dataplane replicas.
+	Replicas int `json:"replicas,omitempty"`
+}
+
+// IngressSpec defines the type of the ingress component in ClusterLink.
+type IngressSpec struct {
+	// +kubebuilder:validation:Enum=none;LoadBalancer;NodePort
+	// +kubebuilder:default=none
+	// Type represents the type of service used to expose the ClusterLink deployment. Supported values: "LoadBalancer","NodePort", "none".
+	Type IngressType `json:"type,omitempty"`
+}
+
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+
+// Operator is the Schema for the ClusterLink operator API used for deploying a ClusterLink instance.
+type Operator struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   OperatorSpec   `json:"spec,omitempty"`
+	Status OperatorStatus `json:"status,omitempty"`
+}
+
+//+kubebuilder:object:root=true
+
+// OperatorList represents a list of Operator objects.
+type OperatorList struct {
```

Comment:
I still don't see the reason for defining this struct if we don't plan to use it.

Reply:
In all the examples I saw there is always a list object.
Also, I tried to remove it and got an error because the listKind always fills with a default value:
https://docs.openshift.com/container-platform/4.15/rest_api/extension_apis/customresourcedefinition-apiextensions-k8s-io-v1.html

Reply:
Indeed I see it is defined by convention. Please ignore my previous comment :)